### PR TITLE
Converted Stateless Schedulers to Classes

### LINKED ISSUE #591: UX Improvements for Schedulers
Could we convert our schedulers into classes, where ssr and state are passed in on __call__, and all other (scheduler-specific) parameters are passed in on init. It's not a great UX that users will have to manually construct lambas to use our schedulers.

See the documentation page here: https://mosaicml-composer--531.com.readthedocs.build/en/531/trainer/optimizers_and_schedulers.html#composerschedulers

Instead, I think it would be cleaner if the user could pass in a `StepScheduler(step_size=['5ep', '25ep'])`, and then the trainer calls `compile_scheduler(step_scheduler, state, ssr)`.

This would also solve this comment which was originally on #463:

To simplify the logic in the training loop and to remove the `functools.partial` in scale scheduler, could `compile_scheduler` take the `ssr` as a parameter?
Then on line 63, we can call the custom scale_pytorch_scheduler logic, and on line 72, we can pass the ssr into the `return scheduler(state, ssr)`

Just not a big fan of functools.partial as they are not typechecked.

_Originally posted by @ravi-mosaicml in https://github.com/mosaicml/composer/pull/463#discussion_r810257320_

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
* Converted all stateless scheduling functions into classes that take scheduler-specific parameters on `__init__`, and have the same `__call__(self, state, ssr = 1.0)` signature.
* Fixed the Exponential LR to support stepwise scheduling, rather than just epoch-wise scheduling. Added a `gamma_period` argument.
* Fixed a bug in the makefile that prevented isort from running properly.

Documentation will be addressed in #398.

Closes #591.

### CODE REVIEW

File: docker/tests/test_pytorch_docker.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env bash
```
Comment (hanlint) [inline docker/tests/test_pytorch_docker.sh:1]: ?

Comment (ravi-mosaicml) [reply docker/tests/test_pytorch_docker.sh:1]: oh whoops didn't mean to commit this file

File: composer/trainer/trainer.py [hunk 2]
```diff
@@ -582,20 +583,21 @@ def __init__(
...
-
-        if scale_schedule_ratio != 1.0:
-            schedulers = tuple(
-                scale_scheduler(scheduler, scale_schedule_ratio) for scheduler in ensure_tuple(schedulers))
-
         if step_schedulers_every_batch is None:
-            step_schedulers_every_batch = any(callable(scheduler) for scheduler in schedulers)
+            # only if all schedulers are ComposerSchedulers, then we can step every batch by default
+            step_schedulers_every_batch = all(
+                not isinstance(scheduler, PytorchScheduler) for scheduler in ensure_tuple(schedulers))
```
Comment (jbloxham) [inline composer/trainer/trainer.py:589]: This is another place where I'm conflicted, here on `any` vs `all`. On the one hand I'd argue that someone use at least one of our `ComposerScheduler` objects should not be surprised by our batchwise stepping. On the other hand, I could see someone being annoyed by having to reconfigure their `PytorchScheduler` objects when they add or remove a `ComposerScheduler`.

File: composer/trainer/trainer.py [hunk 3]
```diff
@@ -582,20 +583,21 @@ def __init__(
...
-        self.state.schedulers = [compile(scheduler, self.state) for scheduler in schedulers]
+        for scheduler in ensure_tuple(schedulers):
+            if isinstance(scheduler, PytorchScheduler):
+                scale_pytorch_scheduler(scheduler, scale_schedule_ratio)
+                self.state.schedulers.append(scheduler)
+            else:  # it's a composer scheduler
+                self.state.schedulers.append(compile_composer_scheduler(scheduler, self.state, scale_schedule_ratio))
+
+        if len(self.state.schedulers) == 0:
+            warnings.warn(f"No scheduler was specified. Defaulting to {repr(schedulers)}")
```
Comment (jbloxham) [inline composer/trainer/trainer.py:600]: Let's make the error more explicit, to say that the learning rate will be constant.

File: composer/core/types.py [hunk 4]
```diff
@@ -166,8 +166,7 @@ def __len__(self) -> int:
 Metrics = Union[Metric, MetricCollection]
 Optimizer = torch.optim.Optimizer
 Optimizers = Many[Optimizer]
-Scheduler = torch.optim.lr_scheduler._LRScheduler
-Schedulers = Many[Scheduler]
+PytorchScheduler = torch.optim.lr_scheduler._LRScheduler
```
Comment (jbloxham) [inline composer/core/types.py:169]: Although nit:
```suggestion
PyTorchScheduler = torch.optim.lr_scheduler._LRScheduler
```

**Review by jbloxham — APPROVED** [review]
On the whole, looks very good and this will make the docs a lot cleaner! I just have a ton of stylistic suggestions.

File: composer/optim/scheduler.py [hunk 5]
```diff
@@ -68,137 +64,110 @@ def _convert_time(time: Union[str, Time[int], Time[float]], state: State, ssr: f
...
-    return factor
+    def __call__(self, state: State, ssr: float = 1.0):
+        milestones = [_convert_time(milestone, state, ssr=ssr) for milestone in self.milestones]
 
+        factor = 1.0
+        for milestone in milestones:
+            if state.timer >= milestone:
+                factor *= self.gamma
 
-def constant_scheduler(state: State,
```
[the remark this answers is not in the document]
Comment (hanlint) [reply composer/optim/scheduler.py:134]: Don't feel strongly, but would also be in favor of keeping it for completeness.

File: composer/optim/scheduler.py [hunk 6]
```diff
@@ -68,137 +64,110 @@ def _convert_time(time: Union[str, Time[int], Time[float]], state: State, ssr: f
...
+    def __init__(self, gamma: float, decay_period: Union[str, Time] = '1ep'):
+        self.gamma = gamma
+        self.decay_period = Time.from_timestring(decay_period) if isinstance(decay_period, str) else decay_period
 
-    return gamma**(current_time.value / ssr)
+    def __call__(self, state: State, ssr: float = 1.0):
+        decay_period = _convert_time(self.decay_period, state, ssr)
+        current_time_in_decay_units = state.timer.get(decay_period.unit) / ssr
 
+        return self.gamma**float(current_time_in_decay_units / decay_period)
```
[the remark this answers is not in the document]
Comment (hanlint) [reply composer/optim/scheduler.py:167]: While equivalent, I think people will naturally look for it by name rather than know they need to configure a StepScheduler with step=1, so in favor of keeping it and removing that extra step.

[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/optim/scheduler.py:167]: Actually, I guess this adjusts the learning rate continuously whereas `StepScheduler` adjusts it discretely, and both behaviors are probably desirable in their own ways.

[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/optim/scheduler.py:167]: Both are of `Time[int]`, and `Time[int]/Time[int]=Time[float]`

**Review by dblalock — APPROVED** [review]
LGTM. I'm tempted to say we should factor out the common code for adding warmup before other schedulers (and maybe chaining together composer schedulers more generally?), but I'd say that's out of scope for this PR.

File: composer/trainer/trainer.py (hunk 2 above)
Comment (dblalock) [reply composer/trainer/trainer.py:589]: Yeah, I suspect that having one scheduler's behavior change depending on the behavior of others will cause someone to get really confused at some point. E.g., "why are my results different when I add in a composer ConstantLR with lr=1.0, which in theory should do nothing?".

Can we just step each scheduler based on its own preference? Might need a LambdaLR subclass with a `def step_every_batch(self) -> bool: return True` method or something (that `compile_composer_scheduler` would use).

Comment (dblalock) [reply composer/trainer/trainer.py:589]: Simple solution would be to just fail fast if different schedulers have different preferences.

File: composer/core/state.py [hunk 7]
```diff
@@ -81,7 +81,7 @@ class State(Serializable):
         precision_context ((precision: Precision) -> ContextManager): Function to produce a context manager to mandate precision.
 
         optimizers (types.Optimizers, optional): The optimizers being used to train the model. Multiple optimizers are not currently supported.
-        schedulers (types.Schedulers, optional): The learning rate schedulers, typically wrapped in :class:`ComposableScheduler`.
+        schedulers (types.PytorchScheduler, optional): The learning rate schedulers.
```
Comment (dblalock) [inline composer/core/state.py:84]:
```suggestion
        schedulers (types.PytorchScheduler, optional): The learning rate scheduler[s].
```
Less implication that there has to be more than one. Also I'm a little confused here because we only allow one scheduler according to the type signature, but the name `schedulers` is plural.

File: composer/trainer/trainer.py (hunk 2 above)
Comment (ravi-mosaicml) [reply composer/trainer/trainer.py:589]: I am leaning towards where it would be more incorrect that a epochwise scheduler is stepped every batch (compared to the loss of precision by stepping every epoch), so I am inclined to require that all schedulers be composer schedulers if this is not set. Added detailed log messages here to explain the behavior.

File: composer/trainer/_scale_schedule.py [hunk 8]
```diff
@@ -82,10 +50,24 @@ def scale_scheduler(scheduler: Union[Scheduler, ComposerSchedulerFn],
...
+        scheduler.step_size = int(scheduler.step_size * ssr)  # type: ignore  -- unknown attribute
+    elif isinstance(scheduler, MultiStepLR):
+        milestones = scheduler.milestones  # type: ignore  -- unknown attribute
+        milestones = Counter([int(ms * ssr) for ms in milestones])
+        scheduler.milestones = milestones  # type: ignore  -- unknown attribute
+    elif isinstance(scheduler, CosineAnnealingLR):
+        scheduler.T_max = int(scheduler.T_max * ssr)  # type: ignore  -- unknown attribute
+    elif isinstance(scheduler, CosineAnnealingWarmRestarts):
+        # TODO: account for warmups
+        scheduler.T_0 = int(scheduler.T_0 * ssr)  # type: ignore  -- unknown attribute
```
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/trainer/_scale_schedule.py:67]: Our new approach is to manually implement schedulers w/ warmup as a single, holistic scheduler.

[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/trainer/_scale_schedule.py:67]: See https://github.com/mosaicml/composer/blob/dev/composer/optim/scheduler.py#L366

File: composer/optim/scheduler.py [hunk 9]
```diff
@@ -68,137 +64,110 @@ def _convert_time(time: Union[str, Time[int], Time[float]], state: State, ssr: f
...
+        del epoch  # unused. Provided by the pytorch LambdaLR
+        return scheduler(state, ssr)
 
     lambda_scheduler = LambdaLR(optimizer, scheduler_fn)
 
     return lambda_scheduler
 
 
-def step_scheduler(state: State, *, ssr: float = 1.0, step_size: Union[str, Time], gamma: float = 0.1) -> float:
+class StepScheduler:
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/optim/scheduler.py:82]: We do not re-document base class docstrings. I purposefully left out the inheritance because Protocols need not be inherited and wanted to (implicitly) ensure that we do not use any `assert isinstance(scheduler, ComposerScheduler)`, since that would fail if the user provided a lambda instead of a ComposerScheduler object.

File: composer/optim/scheduler.py [hunk 10]
```diff
@@ -68,137 +64,110 @@ def _convert_time(time: Union[str, Time[int], Time[float]], state: State, ssr: f
...
         total_time (str or Time): Total time. Default = ``'1dur'``.
     """
 
-    total_time = _convert_time(total_time, state, ssr=ssr)
-    current_time = state.timer.get(total_time.unit)
-    frac_of_total = min(1.0, (current_time / total_time).value)
+    def __init__(self, start_factor: float = 1.0, end_factor: float = 0.0, total_time: Union[str, Time] = '1dur'):
+        self.start_factor = start_factor
+        self.end_factor = end_factor
+        self.total_time = Time.from_timestring(total_time) if isinstance(total_time, str) else total_time
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/optim/scheduler.py:137]: Gonna defer this for now, but agree it could be helpful to have a method like that.

File: composer/optim/scheduler.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jbloxham) [reply composer/optim/scheduler.py:82]: Yeah, I agree that it's important that `ComposerScheduler` stay as a `Protocol`, but the explicit inheritance will make the docs clearer - right now the Sphinx docs describe `CosineAnnealingScheduler`, for instance, as basing `object`, not `ComposerScheduler`.

File: composer/optim/scheduler.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/optim/scheduler.py:137]: See https://github.com/mosaicml/composer/issues/635

File: composer/optim/scheduler.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/optim/scheduler.py:82]: Re: base of object, that's a sphinx issue we need to fix -- it should be possible to hide that (this happens everywhere and it's kinda annoying).
Agree that it would be nice of having a base of ComposerScheduler. I will change that and add a test case.

File: composer/core/state.py (hunk 7 above)
Comment (ravi-mosaicml) [reply composer/core/state.py:84]: See #640.

**Comment by ravi-mosaicml** [comment]
> LGTM. I'm tempted to say we should factor out the common code for adding warmup before other schedulers (and maybe chaining together composer schedulers more generally?), but I'd say that's out of scope for this PR.

Good point; I think down the line we will want to support multiple training periods. Going to create an issue to track that.

**Comment by hanlint** [comment]
> LGTM. I'm tempted to say we should factor out the common code for adding warmup before other schedulers (and maybe chaining together composer schedulers more generally?), but I'd say that's out of scope for this PR.

Actually the above (supporting chaining and warmup) is the complexity that we decided to remove with the schedulers refactor, since there were limited number of use cases, and one can always write your own scheduler for specific examples.

**Comment by ravi-mosaicml** [comment]
> > LGTM. I'm tempted to say we should factor out the common code for adding warmup before other schedulers (and maybe chaining together composer schedulers more generally?), but I'd say that's out of scope for this PR.
> 
> Actually the above (supporting chaining and warmup) is the complexity that we decided to remove with the schedulers refactor, since there were limited number of use cases, and one can always write your own scheduler for specific examples.

KK; I must have missed this in the design document for the stateless schedulers. I presumed we did it this way for now since it was the simplest way to finish up #146, but I do think having a native feature for chained schedulers would be nice.
